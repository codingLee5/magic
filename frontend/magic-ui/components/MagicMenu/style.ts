import { createStyles } from "antd-style"

export const useStyles = createStyles(({ css, isDarkMode, prefixCls, token }) => {
	return {
		menuWrapper: css`
			.${prefixCls}-menu-item-selected.${prefixCls}-menu-item-selected {
				--${prefixCls}-menu-item-selected-bg: transparent;
				--${prefixCls}-menu-item-selected-color: var(--${prefixCls}-menu-item-color);
			}
		`,
		menu: css`
			background-color: transparent;
			--${prefixCls}-menu-active-bar-border-width: 0 !important;


			.${prefixCls}-menu-item {
				gap: 4px;
				font-size: 14px;
				font-style: normal;
				font-weight: 400;

				&.${prefixCls}-menu-item-danger:hover {
					background-color: ${
						isDarkMode
							? token.magicColorUsages.danger.default
							: token.magicColorScales.red[0]
					} !important;
					color: ${isDarkMode ? token.magicColorUsages.white : token.magicColorUsages.danger} !important;
				}
			}

			.${prefixCls}-menu-submenu-title {
				--${prefixCls}-menu-item-selected-color: var(--${prefixCls}-menu-item-color);
				display: flex;
				align-items: center;
				gap: 4px;
				font-size: 14px;
				font-style: normal;
				font-weight: 400;
			}
		`,
	}
})
