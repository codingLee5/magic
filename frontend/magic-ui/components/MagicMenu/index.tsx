import { memo } from "react"
import type { MenuProps } from "antd"
import { Menu } from "antd"
import { useStyles } from "./style"

export type MagicMenuProps = MenuProps

const MagicMenu = memo(function MagicMenu({ rootClassName, className, ...props }: MagicMenuProps) {
	const { styles, cx } = useStyles()

	return (
		<Menu
			rootClassName={cx(styles.menuWrapper, rootClassName)}
			className={cx(styles.menu, className)}
			{...props}
		/>
	)
})

export default MagicMenu
