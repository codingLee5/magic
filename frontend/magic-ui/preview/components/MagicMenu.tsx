import React from "react"
import MagicMenu from "../../components/MagicMenu"
import ComponentDemo from "./Container"

const MenuDemo: React.FC = () => {
	const horizontalItems = [
		{ key: "1", label: "菜单项1" },
		{ key: "2", label: "菜单项2" },
		{ key: "3", label: "菜单项3" },
	]

	const verticalItems = [
		{ key: "1", label: "菜单项1" },
		{ key: "2", label: "菜单项2" },
		{ key: "3", label: "菜单项3" },
	]

	const nestedItems = [
		{ key: "1", label: "菜单项1" },
		{
			key: "2",
			label: "子菜单",
			children: [
				{ key: "2-1", label: "子菜单项1" },
				{ key: "2-2", label: "子菜单项2" },
			],
		},
		{ key: "3", label: "菜单项3" },
	]

	return (
		<div>
			<ComponentDemo title="水平菜单" description="水平方向的菜单" code="mode='horizontal'">
				<MagicMenu
					mode="horizontal"
					items={horizontalItems}
					style={{ borderBottom: "1px solid #f0f0f0" }}
				/>
			</ComponentDemo>

			<ComponentDemo title="垂直菜单" description="垂直方向的菜单" code="mode='vertical'">
				<div style={{ width: 200, border: "1px solid #f0f0f0" }}>
					<MagicMenu mode="vertical" items={verticalItems} />
				</div>
			</ComponentDemo>

			<ComponentDemo title="嵌套菜单" description="支持嵌套子菜单" code="children">
				<div style={{ width: 200, border: "1px solid #f0f0f0" }}>
					<MagicMenu mode="vertical" items={nestedItems} />
				</div>
			</ComponentDemo>

			<ComponentDemo title="事件处理" description="监听菜单点击事件" code="onClick">
				<MagicMenu
					mode="horizontal"
					items={horizontalItems}
					onClick={({ key }) => {
						console.log("点击的菜单项:", key)
					}}
					style={{ borderBottom: "1px solid #f0f0f0" }}
				/>
			</ComponentDemo>

			<ComponentDemo
				title="默认选中"
				description="设置默认选中的菜单项"
				code="defaultSelectedKeys"
			>
				<MagicMenu
					mode="horizontal"
					items={horizontalItems}
					defaultSelectedKeys={["1"]}
					style={{ borderBottom: "1px solid #f0f0f0" }}
				/>
			</ComponentDemo>
		</div>
	)
}

export default MenuDemo
