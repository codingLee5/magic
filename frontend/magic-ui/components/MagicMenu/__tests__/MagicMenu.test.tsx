import { render, screen } from "@testing-library/react"
import MagicThemeProvider from "../../ThemeProvider"
import { describe, it, expect } from "vitest"
import MagicMenu from "../index"

const renderWithTheme = (component: React.ReactElement) =>
	render(<MagicThemeProvider theme="light">{component}</MagicThemeProvider>)

describe("MagicMenu", () => {
	it("应该正常渲染", () => {
		renderWithTheme(
			<MagicMenu
				items={[
					{
						key: "1",
						label: "菜单项1",
					},
				]}
			/>,
		)
		expect(screen.getByText("菜单项1")).toBeInTheDocument()
	})

	// 快照测试
	describe("快照测试", () => {
		it("基础菜单快照", () => {
			const { asFragment } = renderWithTheme(
				<MagicMenu
					items={[
						{ key: "1", label: "菜单项1" },
						{ key: "2", label: "菜单项2" },
					]}
				/>,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("带图标菜单快照", () => {
			const { asFragment } = renderWithTheme(
				<MagicMenu
					items={[
						{ key: "1", label: "菜单项1", icon: <span>🏠</span> },
						{ key: "2", label: "菜单项2", icon: <span>⚙️</span> },
					]}
				/>,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("带子菜单快照", () => {
			const { asFragment } = renderWithTheme(
				<MagicMenu
					items={[
						{
							key: "1",
							label: "父菜单",
							children: [
								{ key: "1-1", label: "子菜单1" },
								{ key: "1-2", label: "子菜单2" },
							],
						},
					]}
				/>,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("选中状态菜单快照", () => {
			const { asFragment } = renderWithTheme(
				<MagicMenu
					selectedKeys={["1"]}
					items={[
						{ key: "1", label: "菜单项1" },
						{ key: "2", label: "菜单项2" },
					]}
				/>,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("垂直菜单快照", () => {
			const { asFragment } = renderWithTheme(
				<MagicMenu
					mode="vertical"
					items={[
						{ key: "1", label: "菜单项1" },
						{ key: "2", label: "菜单项2" },
					]}
				/>,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("带自定义样式菜单快照", () => {
			const { asFragment } = renderWithTheme(
				<MagicMenu
					style={{ width: "200px" }}
					items={[
						{ key: "1", label: "菜单项1" },
						{ key: "2", label: "菜单项2" },
					]}
				/>,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("带类名菜单快照", () => {
			const { asFragment } = renderWithTheme(
				<MagicMenu
					className="custom-menu"
					items={[
						{ key: "1", label: "菜单项1" },
						{ key: "2", label: "菜单项2" },
					]}
				/>,
			)
			expect(asFragment()).toMatchSnapshot()
		})
	})
})
