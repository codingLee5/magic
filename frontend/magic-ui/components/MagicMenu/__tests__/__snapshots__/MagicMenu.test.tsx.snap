// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MagicMenu > т┐ФуЁДТхІУ»Ћ > тъѓуЏ┤УЈютЇЋт┐ФуЁД 1`] = `
<DocumentFragment>
  <ul
    class="magic-menu magic-menu-root magic-menu-vertical magic-menu-light acss-sgfakk acss-1691gak css-dev-only-do-not-override-yotw4n css-var-rk magic-menu-css-var"
    data-menu-list="true"
    role="menu"
    tabindex="0"
  >
    <li
      aria-describedby="test-id"
      class="magic-menu-item magic-menu-item-only-child"
      data-menu-id="rc-menu-uuid-test-1"
      role="menuitem"
      tabindex="-1"
    >
      <span
        class="magic-menu-title-content"
      >
        УЈютЇЋжА╣1
      </span>
    </li>
    <li
      aria-describedby="test-id"
      class="magic-menu-item magic-menu-item-only-child"
      data-menu-id="rc-menu-uuid-test-2"
      role="menuitem"
      tabindex="-1"
    >
      <span
        class="magic-menu-title-content"
      >
        УЈютЇЋжА╣2
      </span>
    </li>
  </ul>
  <div
    aria-hidden="true"
    style="display: none;"
  />
</DocumentFragment>
`;

exports[`MagicMenu > т┐ФуЁДТхІУ»Ћ > тЪ║уАђУЈютЇЋт┐ФуЁД 1`] = `
<DocumentFragment>
  <ul
    class="magic-menu magic-menu-root magic-menu-vertical magic-menu-light acss-sgfakk acss-1691gak css-dev-only-do-not-override-yotw4n css-var-r3 magic-menu-css-var"
    data-menu-list="true"
    role="menu"
    tabindex="0"
  >
    <li
      aria-describedby="test-id"
      class="magic-menu-item magic-menu-item-only-child"
      data-menu-id="rc-menu-uuid-test-1"
      role="menuitem"
      tabindex="-1"
    >
      <span
        class="magic-menu-title-content"
      >
        УЈютЇЋжА╣1
      </span>
    </li>
    <li
      aria-describedby="test-id"
      class="magic-menu-item magic-menu-item-only-child"
      data-menu-id="rc-menu-uuid-test-2"
      role="menuitem"
      tabindex="-1"
    >
      <span
        class="magic-menu-title-content"
      >
        УЈютЇЋжА╣2
      </span>
    </li>
  </ul>
  <div
    aria-hidden="true"
    style="display: none;"
  />
</DocumentFragment>
`;

exports[`MagicMenu > т┐ФуЁДТхІУ»Ћ > тИдтЏЙТаЄУЈютЇЋт┐ФуЁД 1`] = `
<DocumentFragment>
  <ul
    class="magic-menu magic-menu-root magic-menu-vertical magic-menu-light acss-sgfakk acss-1691gak css-dev-only-do-not-override-yotw4n css-var-r8 magic-menu-css-var"
    data-menu-list="true"
    role="menu"
    tabindex="0"
  >
    <li
      aria-describedby="test-id"
      class="magic-menu-item"
      data-menu-id="rc-menu-uuid-test-1"
      role="menuitem"
      tabindex="-1"
    >
      <span
        class="magic-menu-item-icon"
      >
        ­ЪЈа
      </span>
      <span
        class="magic-menu-title-content"
      >
        УЈютЇЋжА╣1
      </span>
    </li>
    <li
      aria-describedby="test-id"
      class="magic-menu-item"
      data-menu-id="rc-menu-uuid-test-2"
      role="menuitem"
      tabindex="-1"
    >
      <span
        class="magic-menu-item-icon"
      >
        РџЎ№ИЈ
      </span>
      <span
        class="magic-menu-title-content"
      >
        УЈютЇЋжА╣2
      </span>
    </li>
  </ul>
  <div
    aria-hidden="true"
    style="display: none;"
  />
</DocumentFragment>
`;

exports[`MagicMenu > т┐ФуЁДТхІУ»Ћ > тИдтГљУЈютЇЋт┐ФуЁД 1`] = `
<DocumentFragment>
  <ul
    class="magic-menu magic-menu-root magic-menu-vertical magic-menu-light acss-sgfakk acss-1691gak css-dev-only-do-not-override-yotw4n css-var-rd magic-menu-css-var"
    data-menu-list="true"
    role="menu"
    tabindex="0"
  >
    <li
      class="magic-menu-submenu magic-menu-submenu-vertical"
      role="none"
    >
      <div
        aria-controls="rc-menu-uuid-test-1-popup"
        aria-expanded="false"
        aria-haspopup="true"
        class="magic-menu-submenu-title"
        data-menu-id="rc-menu-uuid-test-1"
        role="menuitem"
        tabindex="-1"
      >
        <span
          class="magic-menu-title-content"
        >
          уѕХУЈютЇЋ
        </span>
        <i
          class="magic-menu-submenu-arrow"
        />
      </div>
    </li>
  </ul>
  <div
    aria-hidden="true"
    style="display: none;"
  />
</DocumentFragment>
`;

exports[`MagicMenu > т┐ФуЁДТхІУ»Ћ > тИду▒╗тљЇУЈютЇЋт┐ФуЁД 1`] = `
<DocumentFragment>
  <ul
    class="magic-menu magic-menu-root magic-menu-vertical magic-menu-light acss-sgfakk custom-menu acss-1691gak css-dev-only-do-not-override-yotw4n css-var-ru magic-menu-css-var"
    data-menu-list="true"
    role="menu"
    tabindex="0"
  >
    <li
      aria-describedby="test-id"
      class="magic-menu-item magic-menu-item-only-child"
      data-menu-id="rc-menu-uuid-test-1"
      role="menuitem"
      tabindex="-1"
    >
      <span
        class="magic-menu-title-content"
      >
        УЈютЇЋжА╣1
      </span>
    </li>
    <li
      aria-describedby="test-id"
      class="magic-menu-item magic-menu-item-only-child"
      data-menu-id="rc-menu-uuid-test-2"
      role="menuitem"
      tabindex="-1"
    >
      <span
        class="magic-menu-title-content"
      >
        УЈютЇЋжА╣2
      </span>
    </li>
  </ul>
  <div
    aria-hidden="true"
    style="display: none;"
  />
</DocumentFragment>
`;

exports[`MagicMenu > т┐ФуЁДТхІУ»Ћ > тИдУЄфт«џС╣ЅТаит╝ЈУЈютЇЋт┐ФуЁД 1`] = `
<DocumentFragment>
  <ul
    class="magic-menu magic-menu-root magic-menu-vertical magic-menu-light acss-sgfakk acss-1691gak css-dev-only-do-not-override-yotw4n css-var-rp magic-menu-css-var"
    data-menu-list="true"
    role="menu"
    style="width: 200px;"
    tabindex="0"
  >
    <li
      aria-describedby="test-id"
      class="magic-menu-item magic-menu-item-only-child"
      data-menu-id="rc-menu-uuid-test-1"
      role="menuitem"
      tabindex="-1"
    >
      <span
        class="magic-menu-title-content"
      >
        УЈютЇЋжА╣1
      </span>
    </li>
    <li
      aria-describedby="test-id"
      class="magic-menu-item magic-menu-item-only-child"
      data-menu-id="rc-menu-uuid-test-2"
      role="menuitem"
      tabindex="-1"
    >
      <span
        class="magic-menu-title-content"
      >
        УЈютЇЋжА╣2
      </span>
    </li>
  </ul>
  <div
    aria-hidden="true"
    style="display: none;"
  />
</DocumentFragment>
`;

exports[`MagicMenu > т┐ФуЁДТхІУ»Ћ > жђЅСИГуіХТђЂУЈютЇЋт┐ФуЁД 1`] = `
<DocumentFragment>
  <ul
    class="magic-menu magic-menu-root magic-menu-vertical magic-menu-light acss-sgfakk acss-1691gak css-dev-only-do-not-override-yotw4n css-var-rf magic-menu-css-var"
    data-menu-list="true"
    role="menu"
    tabindex="0"
  >
    <li
      aria-describedby="test-id"
      class="magic-menu-item magic-menu-item-selected magic-menu-item-only-child"
      data-menu-id="rc-menu-uuid-test-1"
      role="menuitem"
      tabindex="-1"
    >
      <span
        class="magic-menu-title-content"
      >
        УЈютЇЋжА╣1
      </span>
    </li>
    <li
      aria-describedby="test-id"
      class="magic-menu-item magic-menu-item-only-child"
      data-menu-id="rc-menu-uuid-test-2"
      role="menuitem"
      tabindex="-1"
    >
      <span
        class="magic-menu-title-content"
      >
        УЈютЇЋжА╣2
      </span>
    </li>
  </ul>
  <div
    aria-hidden="true"
    style="display: none;"
  />
</DocumentFragment>
`;
